flake.nix

```diff
@@ -0,0 +1,149 @@
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs?rev=ea5234e7073d5f44728c499192544a84244bf35a";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils }:
+    flake-utils.lib.eachSystem [ "x86_64-linux" "x86_64-windows" "x86_64-darwin" "aarch64-darwin" ] (targetSystem:
+      let
+        # This needs to be made dynamic using something like builtins.currentSystem
+        buildSystem = "x86_64-linux";
+
+        systems = {
+          "x86_64-linux" = [ "linux" "x64" ];
+          "x86_64-windows" = [ "win32" "x64" ];
+          "x86_64-darwin" = [ "darwin" "x64" ];
+          "aarch64-darwin" = [ "darwin" "arm64" ];
+        };
+
+        platform = builtins.elemAt systems.${targetSystem} 0;
+        arch = builtins.elemAt systems.${targetSystem} 1;
+
+        pkgs = import nixpkgs {
+          system = buildSystem;
+        };
+
+        utils = pkgs.callPackage ./utils.nix {};
+
+        commitHash = toString (self.rev or self.dirtyRev);
+        npmDepsHash = builtins.readFile ./npmDepsHash;
+
+        polykey_cli-src = pkgs.buildNpmPackage {
+          inherit npmDepsHash;
+          pname = utils.packageName;
+          version = utils.packageVersion;
+          src = utils.src;
+          COMMIT_HASH = commitHash;
+          GIT_DIR = if commitHash != null then null else utils.dotGit;
+          postInstall = ''
+            mv "$packageOut"/build/build.json "$out"/build.json;
+            rm -rf \
+              "$packageOut"/build \
+              "$packageOut"/src \
+              "$packageOut"/.env.example \
+              "$packageOut"/images \
+              "$packageOut"/scripts \
+              "$packageOut"/tsconfig.build.json \
+              "$packageOut"/tsconfig.json \
+              "$packageOut"/LICENSE \
+              "$packageOut"/ADDITIONAL_TERMS \
+              "$packageOut"/README.md;
+          '';
+        };
+
+        polykey_cli-release = pkgs.buildNpmPackage {
+          inherit npmDepsHash;
+          name = "${utils.packageName}-${utils.packageVersion}-${platform}-${arch}";
+          src = utils.src;
+          PKG_CACHE_PATH = utils.pkgCachePath;
+          PKG_IGNORE_TAG = 1;
+          COMMIT_HASH = commitHash;
+          GIT_DIR = if commitHash != null then null else utils.dotGit;
+          postBuild = ''
+            npm run pkg -- \
+              --output=out \
+              --bin=dist/polykey.js \
+              --node-version=${utils.nodeVersion} \
+              --platform=${platform} \
+              --arch=${arch}
+          '';
+          installPhase = ''
+            cp ${if platform == "win32" then "out.exe" else "out"} $out
+          '';
+          dontFixup = true;
+        };
+
+        buildJSON = builtins.fromJSON (builtins.readFile "${polykey_cli-src}/build.json");
+
+        dockerImage = pkgs.dockerTools.buildImage {
+          name = polykey_cli-src.name;
+          copyToRoot = [ polykey_cli-src ];
```

Comment:
Hmm I'm not sure if I like this. `src` does not make sense here.

Reply:
To explain, I put two derivations of Polykey-CLI inside the flake, `polykey_cli-src` and `polykey_cli-release`. One of them contains the entire source tree, and is executed with Node automatically when running the output, which itself is a symlink to a script that will run `polykey.js` with Node.

I kept this like this because this is what the original `release.nix` was doing, it was importing the `default.nix` which used an, almost identical to what I have now, build process using `buildNpmPackage`.

I don't know if this should be changed or not, it seems to work perfectly fine as is right now.

Reply:
So the idea of `src` isn't quite right. I consider that to be an "application" build output or the default build output.

The pkg built ones, should then be suffixed with the relevant output.

So let's make up a spec.

`polykey-cli` is the default.

`polykey-cli-pkg` is the pkg related stuff.

So then `polykey-cli-pkg-win32` should be...?

Think about it that way. Look at how we are naming things across the board, prefix is always the "bigger picture".

BTW maybe this wasn't documented anywhere, but we follow SOV naming style. Subject Object Verb. Or in this case, Subject Object.

Reply:
I renamed them to what you suggested in 1d7af25. Its a lot clearer now.